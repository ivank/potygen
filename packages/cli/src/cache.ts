import { existsSync, readFileSync, writeFileSync, statSync, mkdirSync } from 'fs';
import { dirname } from 'path';
import { createHash } from 'crypto';

export class CachedFileParser<TResult> {
  constructor(public store = new Map<string, TResult>(), public process: (content: string, path: string) => TResult) {}

  public get(path: string): TResult {
    const content = readFileSync(path, 'utf-8');
    const key = createHash('md5').update(content).digest('hex');
    const cachedResult = this.store.get(key);
    if (cachedResult) {
      return cachedResult;
    } else {
      const result = this.process(content, path);
      this.store.set(key, result);
      return result;
    }
  }
}

export class CachedProcessParser<TResult> {
  constructor(public store = new Map<string, TResult>(), public process: (content: string, path: string) => TResult) {}

  public get(path: string): TResult {
    const content = readFileSync(path, 'utf-8');
    const key = createHash('md5').update(content).digest('hex');
    const cachedResult = this.store.get(key);
    if (cachedResult) {
      return cachedResult;
    } else {
      const result = this.process(content, path);
      this.store.set(key, result);
      return result;
    }
  }
}

export class CacheStore {
  public store = new Map<string, number>();

  constructor(public fileName: string, public enabled: boolean = false, public cacheClear: boolean = false) {
    if (this.enabled && !this.cacheClear) {
      if (existsSync(fileName)) {
        this.store = new Map(Object.entries(JSON.parse(readFileSync(fileName, 'utf-8'))));
      }
    }
  }

  public shouldParseFile(path: string) {
    if (!this.enabled) {
      return true;
    } else {
      const mtime = statSync(path).mtime.getTime();
      const currentMtime = this.store.get(path);
      if (currentMtime && mtime === currentMtime) {
        return false;
      }
    }
    return true;
  }

  public cacheFile(path: string) {
    const mtime = statSync(path).mtime.getTime();
    this.store.set(path, mtime);
  }

  public save() {
    if (this.enabled) {
      mkdirSync(dirname(this.fileName), { recursive: true });
      writeFileSync(this.fileName, JSON.stringify(Object.fromEntries(this.store)), 'utf-8');
    }
  }
}
