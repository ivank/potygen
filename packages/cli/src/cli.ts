import { Client } from 'pg';
import { pipeline } from 'stream';
import { promisify } from 'util';
import { SqlRead, QueryLoader } from './traverse';

const asyncPipeline = promisify(pipeline);

const main = async () => {
  const db = new Client({
    database: 'schemes',
    user: 'schemes',
    password: 'dev-pass',
    host: '127.0.0.1',
    port: 5432,
    ssl: false,
  });
  try {
    await db.connect();

    const sqls = new SqlRead('src/**/*.query.ts', '../../../schemes/packages/api');
    const sink = new QueryLoader(
      db,
      '../../../schemes/packages/api',
      '{{root}}/src/queries/__psql-ts__/{{name}}.queries.ts',
    );

    await asyncPipeline(sqls, sink);
  } catch (error) {
    console.log(String(error));
  } finally {
    await db.end();
  }
};

main();
