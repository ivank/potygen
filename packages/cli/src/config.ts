import { Record, Optional, Static, String, Boolean } from 'runtypes';

export const Config = Record({
  files: Optional(String),
  watch: Optional(Boolean),
  root: Optional(String),
  connection: Optional(String),
  template: Optional(String),
  verbose: Optional(Boolean),
  silent: Optional(Boolean),
  typePrefix: Optional(String),
  preload: Optional(Boolean),
  cache: Optional(Boolean),
  cacheFile: Optional(String),
  cacheClear: Optional(Boolean),
});

export type ConfigType = Static<typeof Config>;
export type FullConfigType = Required<ConfigType>;

export const toConfig = (config: unknown): FullConfigType => ({
  files: '**/*.sql',
  cacheFile: '.cache/potygen.cache',
  root: process.cwd(),
  preload: false,
  template: '{{dir}}/{{name}}.queries.ts',
  connection: 'postgres://localhost:5432/db',
  watch: false,
  verbose: false,
  silent: false,
  cache: false,
  cacheClear: false,
  typePrefix: '',
  ...Config.check(config),
});
