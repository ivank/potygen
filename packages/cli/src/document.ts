import { document, printDocument, DocumentContext, Type, withIdentifier } from '@ovotech/ts-compose';
import { TypeNode } from 'typescript';
import { isLoadedArrayType, isLoadedUnionType, isLoadedLiteralType, isLoadedValuesPick } from './guards';
import { LoadedQuery, LoadedValuesPick, LoadedType } from './types';

const toPropertyType = (type: LoadedType | LoadedValuesPick): TypeNode => {
  if (isLoadedValuesPick(type)) {
    return Type.Array(
      Type.TypeLiteral({
        props: type.items.map((item) => Type.Prop({ name: item.name, type: toPropertyType(item.value) })),
      }),
    );
  } else if (isLoadedArrayType(type)) {
    return Type.Array(toPropertyType(type.items));
  } else if (isLoadedUnionType(type)) {
    return Type.Union(type.items.map(toPropertyType));
  } else if (isLoadedLiteralType(type)) {
    return Type.Literal(type.value);
  } else {
    switch (type.type) {
      case 'Date':
        return Type.Referance('Date');
      case 'boolean':
        return Type.Boolean;
      case 'json':
        return Type.Unknown;
      case 'null':
        return Type.Null;
      case 'number':
        return Type.Number;
      case 'string':
        return Type.String;
      case 'unknown':
        return Type.Unknown;
    }
  }
};

export const toQueryTypescriptTypes = (context: DocumentContext, query: LoadedQuery) => {
  const paramsContext = withIdentifier(
    context,
    Type.Interface({
      name: 'QueryParams',
      isExport: true,
      props: query.params.map(({ name, type }) =>
        Type.Prop({ name, type: toPropertyType(type), isOptional: type.optional }),
      ),
    }),
  );

  const resultContext = withIdentifier(
    paramsContext,
    Type.Interface({
      name: 'QueryResult',
      isExport: true,
      props: query.result.map(({ name, type }) =>
        Type.Prop({ name, type: toPropertyType(type), isOptional: type.optional }),
      ),
    }),
  );

  return document(
    resultContext,
    Type.Interface({
      name: 'Query',
      isExport: true,
      props: [
        Type.Prop({ name: 'params', type: Type.Referance('QueryParams') }),
        Type.Prop({ name: 'result', type: Type.Referance('QueryResult') }),
      ],
    }),
  );
};

export const toQueryTypescript = (loadedQuery: LoadedQuery): string => {
  return printDocument(toQueryTypescriptTypes({}, loadedQuery));
};
