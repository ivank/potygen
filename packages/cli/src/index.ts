export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
} from './types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
} from './guards';

export { toLoadedContext, loadQueryInterfacesData, toLoadedQueryInterface, loadData, extractDataSources } from './load';

export { Config, ConfigType, FullConfigType, toConfig } from './config';

export { emitLoadedFile, toTypeSource } from './emit';
export { glob } from './glob';
