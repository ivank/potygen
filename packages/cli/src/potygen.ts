import { Logger } from '@potygen/potygen';
import { existsSync, readFileSync } from 'fs';
import { inspect, promisify } from 'util';
import { Client } from 'pg';
import { pipeline } from 'stream';
import { Command, createCommand } from 'commander';
import { QueryLoader, SqlRead } from './traverse';
import { ConfigType, toConfig } from './config';

const asyncPipeline = promisify(pipeline);

enum LogLevel {
  error,
  info,
  debug,
}

class LogLevelConsole implements Logger {
  constructor(public level: LogLevel) {}

  error(...args: any[]): void {
    if (this.level >= LogLevel.error) {
      console.info(...args);
    }
  }

  info(...args: any[]): void {
    if (this.level >= LogLevel.info) {
      console.info(...args);
    }
  }

  debug(...args: any[]): void {
    if (this.level >= LogLevel.debug) {
      console.debug(...args);
    }
  }
}

export const potygen = (overwriteLogger?: Logger): Command =>
  createCommand('potygen')
    .description('Convert postgres query files inty typescript types')
    .version('0.5.11')
    .option('-c, --config <config>', 'A configuration file to load', 'potygen.config.json')
    .option('-f, --files <files>', 'A glob pattern to search files by (default: "**/*.sql")')
    .option('-w, --watch', 'Watch for file changes and update live')
    .option('-v, --verbose', 'Show verbose logs')
    .option('-s, --silent', 'Only show error logs')
    .option('-p, --typePrefix <typePrefix>', 'Prefix generated types')
    .option('-r, --root <root>', `Set the root directory (default: ${process.cwd()})`)
    .option(
      '-n, --connection <connection>',
      'Connection to the postgres database. URI (default: "postgres://localhost:5432/db")',
    )
    .option(
      '-t, --template <template>',
      'A template of the path, where to generate the typescript type files. The parameters are the response from node\'s path.parse function (default: "{{dir}}/{{name}}.queries.ts")',
    )
    .action(async (options: ConfigType & { config: string }) => {
      const { config, ...rest } = options;
      const { root, connection, watch, files, template, verbose, silent, typePrefix } = toConfig({
        ...(existsSync(config) ? JSON.parse(readFileSync(config, 'utf-8')) : {}),
        ...rest,
      });

      const logger =
        overwriteLogger ?? new LogLevelConsole(verbose ? LogLevel.debug : silent ? LogLevel.error : LogLevel.info);

      logger.debug(`Potygen Config: ${inspect({ root, connection, watch, files, template })}`);

      const db = new Client(connection);
      await db.connect();
      try {
        const sqls = new SqlRead({ path: files, root, watch, logger });
        const sink = new QueryLoader({ db, root, template, logger, typePrefix });

        logger.info(`Potygen started processing ("${files}", watch: ${watch})`);

        await asyncPipeline(sqls, sink);
      } catch (error) {
        if (process.env.POTYGEN_DEBUG && error instanceof Error) {
          logger.error(error.stack);
        } else {
          logger.error(String(error));
        }
      } finally {
        await db.end();
      }
    });
