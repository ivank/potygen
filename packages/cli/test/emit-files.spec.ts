import { parser, toQueryInterface, loadQueryInterfacesData, toLoadedQueryInterface } from '@potygen/potygen';
import { Client } from 'pg';
import { createPrinter, NewLineKind } from 'typescript';
import { toTypeSource } from '../src';
import { sqlFiles, testDb, withParserErrors } from './helpers';

let db: Client;

describe('Load Files', () => {
  beforeAll(async () => {
    db = testDb();
    await db.connect();
  });

  afterAll(() => db.end());

  it.each(sqlFiles())(
    'Should convert complex sql %s',
    (path, content) =>
      withParserErrors(async () => {
        const logger = { info: jest.fn(), error: jest.fn(), debug: jest.fn() };
        const printer = createPrinter({ newLine: NewLineKind.LineFeed });
        const { ast } = parser(content);
        const queryInterface = toQueryInterface(ast);

        const data = await loadQueryInterfacesData({ db, logger }, [queryInterface], []);
        const loadedQuery = toLoadedQueryInterface(data)(queryInterface);
        const source = toTypeSource({ type: 'sql', path: path, content, queryInterface, loadedQuery });
        expect(printer.printFile(source)).toMatchSnapshot(path);
      }),
    10000,
  );
});
