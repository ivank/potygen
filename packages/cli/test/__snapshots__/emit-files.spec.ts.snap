// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load Files Should convert complex sql account-levelisation.update.query.sql: account-levelisation.update.query.sql 1`] = `
"export interface Params {
    isAccepted?: boolean;
    id?: number;
}
export interface Result {
    id: number;
    isAccepted: boolean;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql account-levelisation.upsert.query.sql: account-levelisation.upsert.query.sql 1`] = `
"export interface Params {
    levelisationId?: number;
    start?: Date;
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql account-levelisations.query.sql: account-levelisations.query.sql 1`] = `
"import { Json } from \\"@potygen/potygen\\";
export interface Params {
    q?: string;
    resolvedPostlev?: \\"\\" | boolean;
    state?: (\\"Pending\\" | \\"Done\\");
    levelisationId?: number;
    ids?: number[];
    sortField?: \\"totalPayment\\" | \\"vatPayment\\" | \\"generationPayment\\" | \\"exportPayment\\" | \\"generationEnergy\\" | \\"exportEnergy\\" | \\"cfrFitId\\" | \\"state\\" | \\"isAccepted\\" | \\"error\\" | \\"exportType\\";
    sortOrder?: \\"DESC\\" | \\"ASC\\";
    limit?: number;
    offset?: number;
}
export interface Result<TGenerationPeriods = unknown, TExportPeriods = unknown, TErrorParams = unknown> {
    id: number;
    levelisationId: number;
    accountId: number;
    installationId: number;
    state: \\"Pending\\" | \\"Done\\";
    isAccepted: boolean;
    generationStartReadOn?: Date;
    generationStartReadValue?: number;
    generationEndReadOn?: Date;
    generationEndReadValue?: number;
    generationPercentageSplit?: number;
    generationPayment?: number;
    generationEnergy?: number;
    exportStartReadOn?: Date;
    exportStartReadValue?: number;
    exportEndReadOn?: Date;
    exportEndReadValue?: number;
    exportPercentageSplit?: number;
    exportPayment?: number;
    exportEnergy?: number;
    exportType?: \\"Deemed\\" | \\"Metered Export\\" | \\"Off Grid\\" | \\"PPA\\";
    technologyType?: \\"PV\\" | \\"H\\" | \\"W\\" | \\"AD\\" | \\"CHP\\";
    totalPayment?: number;
    vatPayment?: number;
    generationPeriods?: Json<TGenerationPeriods>;
    exportPeriods?: Json<TExportPeriods>;
    errorCode?: string;
    errorParams?: Json<TErrorParams>;
    cfrFitId: string;
    resolvedPostlevId?: number;
    isBacsPaymentsSent?: boolean;
    isChequePaymentsSent?: boolean;
}
export interface Query<TGenerationPeriods = unknown, TExportPeriods = unknown, TErrorParams = unknown> {
    params: Params;
    result: Result<TGenerationPeriods, TExportPeriods, TErrorParams>;
}
"
`;

exports[`Load Files Should convert complex sql account-levelisations-bulk-close-accounts.query.sql: account-levelisations-bulk-close-accounts.query.sql 1`] = `
"export interface Params {
    ids?: number[];
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql account-levelisations-set-bulk-update.query.sql: account-levelisations-set-bulk-update.query.sql 1`] = `
"export interface Params {
    ids?: number[];
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql accounts.list.query.sql: accounts.list.query.sql 1`] = `
"export interface Params {
    q?: string;
    ids?: number[];
}
export interface Result {
    total: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql accounts.query.sql: accounts.query.sql 1`] = `
"export interface Params {
    q?: string;
    ids?: number[];
    limit?: number;
    offset?: number;
}
export interface Result {
    id: number;
    state: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
    startOn?: Date;
    firstName: string;
    lastName: string;
    email?: string;
    address: string;
    fitIds: string[];
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql accounts.sql: accounts.sql 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql active-meters-with-last-read.query.sql: active-meters-with-last-read.query.sql 1`] = `
"export interface Params {
    intervalStart?: Date;
    intervalEnd?: Date;
    accountId?: number;
    meterId?: number;
}
export interface Result {
    id: number;
    msn: string;
    mpan: string;
    make: string;
    model: string;
    type: \\"Export\\" | \\"Generation\\";
    currentPeriodRead?: {
        id: number;
        value: number;
        dateOn: string;
        submittedAt?: string;
        reason?: string;
        type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    };
    previousPeriodRead?: {
        id: number;
        value: number;
        dateOn: string;
        submittedAt?: string;
        reason?: string;
        type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    };
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql active-meters-wth-reads.query.sql: active-meters-wth-reads.query.sql 1`] = `
"export interface Params {
    intervalStart: Date;
    intervalEnd: Date;
    accountId: number;
    meterId?: number;
}
export interface Result {
    id: number;
    msn: string;
    mpan: string;
    make: string;
    model: string;
    type: \\"Export\\" | \\"Generation\\";
    currentPeriodRead?: {
        id: number;
        value: number;
        dateOn: string;
        submittedAt?: string;
        reason?: string;
        type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    };
    previousPeriodRead?: {
        id: number;
        value: number;
        dateOn: string;
        submittedAt?: string;
        reason?: string;
        type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    };
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql active-reads.view.sql: active-reads.view.sql 1`] = `
"import { Json } from \\"@potygen/potygen\\";
export interface Params {
}
export interface Result<THistory = unknown> {
    id: number;
    meter_id: number;
    date_on: Date;
    value: number;
    type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    reason?: string;
    created_at: Date;
    updated_at?: Date;
    /**
     * sql-fit: MeterReading.MeterReadingId
     */
    source_system_id?: number;
    checked: boolean;
    deleted_at?: Date;
    submitted_at?: Date;
    history: Json<THistory>;
    overwritten_at?: Date;
}
export interface Query<THistory = unknown> {
    params: Params;
    result: Result<THistory>;
}
"
`;

exports[`Load Files Should convert complex sql bacs.query.sql: bacs.query.sql 1`] = `
"export interface Params {
    quarter?: string;
    perPage?: string;
    offset?: string;
}
export interface Result {
    fitId: string;
    name?: string;
    amount: number;
    quarter?: string;
    bankDetails: string;
    addressLine1?: string;
    addressLine2?: string;
    addressLine3?: string;
    postcode: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql contracts.sql: contracts.sql 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql csa-account-by-id.query.sql: csa-account-by-id.query.sql 1`] = `
"export interface Params {
    id?: number;
}
export interface Result {
    id: number;
    state: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
    contacts: {
        id: number;
        name: string;
        email?: string;
        phone?: string;
        address: {
            id: number;
            addressLine1?: string;
            addressLine2?: string;
            addressLine3?: string;
            city?: string;
            country: string;
            county?: string;
            postCode: string;
        };
        isPrimaryContact: boolean;
        isNominatedRecipient: boolean;
    }[];
    paymentDetails: {
        beneficiaryAccountNumber: string;
        beneficiaryName?: string;
        beneficiarySortCode?: string;
    };
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql cte.query.sql: cte.query.sql 1`] = `
"export interface Params {
}
export interface Result {
    region: string;
    product: string;
    product_units: number;
    product_sales: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql cte-levelisation-details.query.sql: cte-levelisation-details.query.sql 1`] = `
"export interface Params {
    id?: number;
}
export interface Result {
    id: number;
    totalCount?: number;
    fitTotalExportPayment?: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql delete-accounts.query.sql: delete-accounts.query.sql 1`] = `
"export interface Params {
    type?: \\"FIT\\" | \\"SEG\\";
    state?: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
}
export interface Result {
    id: number;
    sourceSystemId?: number;
    customerId: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql generation-read-data.query.sql: generation-read-data.query.sql 1`] = `
"export interface Params {
    installationId?: number;
    exportDateOn?: Date;
}
export interface Result {
    dateOn: Date;
    value: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql insert.process.query.sql: insert.process.query.sql 1`] = `
"import { Json } from \\"@potygen/potygen\\";
export interface Params<TParamItemsData = unknown> {
    items: {
        processId: number;
        idempotencyKey?: string;
        data: TParamItemsData;
        accountId: number;
    }[];
}
export interface Result<TData = unknown> {
    id: number;
    data: Json<TData>;
}
export interface Query<TParamItemsData = unknown, TData = unknown> {
    params: Params<TParamItemsData>;
    result: Result<TData>;
}
"
`;

exports[`Load Files Should convert complex sql installation-meter-update.query.sql: installation-meter-update.query.sql 1`] = `
"export interface Params {
    removalDate?: Date;
    meterId?: number;
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql installations.sql: installations.sql 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql levelisation-customer-data.query.sql: levelisation-customer-data.query.sql 1`] = `
"export interface Params {
    accountLevelisationId: number;
}
export interface Result {
    interval: {
        start: string;
        end: string;
    };
    levelisationQuarter: string;
    accountId: number;
    contact?: {
        title?: \\"Mrs\\" | \\"Miss\\" | \\"Dr\\" | \\"Ms\\" | \\"Mr\\";
        firstName: string;
        lastName: string;
        email?: string;
        address: {
            city?: string;
            postcode: string;
            line1?: string;
            line2?: string;
            line3?: string;
        };
    };
    installation?: {
        installationId: number;
        technologyType: \\"PV\\" | \\"H\\" | \\"W\\" | \\"AD\\" | \\"CHP\\";
        meters: {
            id: number;
            startOn?: string;
            endOn?: string;
            type: \\"Export\\" | \\"Generation\\";
            reads: {
                value: number;
                dateOn: string;
                type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
                id: number;
            }[];
        }[];
        contracts: {
            schemeType: \\"FIT\\" | \\"SEG\\";
            schemeAccountReference: string;
            capacity: number;
            generationTariffId?: number;
            exportTariffId?: number;
            startOn: string;
            endOn?: string;
            exportType?: \\"Deemed\\" | \\"Metered Export\\" | \\"Off Grid\\" | \\"PPA\\";
            exportPercentageSplit?: number;
            generationPercentageSplit?: number;
            generationRates: {
                id: number;
                startOn: string;
                endOn: string;
                rate: number;
            }[];
            exportRates: {
                id: number;
                startOn: string;
                endOn: string;
                rate: number;
            }[];
        }[];
    };
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql levelisation-details.query.sql: levelisation-details.query.sql 1`] = `
"export interface Params {
    id?: number;
}
export interface Result {
    id: number;
    quarter: string;
    startOn: Date;
    endOn: Date;
    totalCount: number;
    pendingCount: number;
    errorCount: number;
    successCount: number;
    backsSent: number;
    chequeSent: number;
    totalGenerationPayment: number;
    totalExportPayment: number;
    totalGenerationEnergy: number;
    totalExportEnergy: number;
    totalMeteredExportEnergy: number;
    totalMeteredExportPayment: number;
    totalDeemedExportEnergy: number;
    totalDeemedExportPayment: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql levelisation-individual-reads.list.query.sql: levelisation-individual-reads.list.query.sql 1`] = `
"export interface Params {
    id?: number;
    limit: number;
    offset: number;
}
export interface Result {
    id: number;
    dateOn: Date;
    value: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql levelisations.list.query.sql: levelisations.list.query.sql 1`] = `
"export interface Params {
    q?: string;
    ids?: number[];
    sortField?: \\"quarter\\";
    sortOrder?: \\"DESC\\" | \\"ASC\\";
    limit?: number;
    offset?: number;
}
export interface Result {
    id: number;
    quarter: string;
    startOn: Date;
    endOn: Date;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql meter-create.query.sql: meter-create.query.sql 1`] = `
"export interface Params {
    id?: number;
    msn?: string;
    make?: string;
    model?: string;
    mpan?: string;
    createdAt?: Date;
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql meter-reads.sql: meter-reads.sql 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql overwritten.query.sql: overwritten.query.sql 1`] = `
"export interface Params {
    currentDate?: Date;
    meterId?: number;
    intervalStart?: Date;
    intervalEnd?: Date;
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql process-batch.query.sql: process-batch.query.sql 1`] = `
"export interface Params {
    fitIds?: string[];
}
export interface Result {
    cfrFiTId?: string;
    email?: string;
    contactName: string;
    company?: string;
    line1?: string;
    line2: string;
    town?: string;
    postcode: string;
    county?: string;
    country: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql rates-by-tarrif-ids.query.sql: rates-by-tarrif-ids.query.sql 1`] = `
"export interface Params {
    ids: number[][];
}
export interface Result {
    tariffId: number;
    rate: number;
    startOn: Date;
    tariffCode: string;
    tariffType: \\"Export\\" | \\"Generation\\" | \\"SEG\\";
    endOn?: Date;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql read-create.query.sql: read-create.query.sql 1`] = `
"export interface Params {
    meterId?: number;
    reason?: string;
    type?: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    dateOn?: Date;
    submittedAt?: Date;
    value?: number;
}
export interface Result {
    id: number;
    value: number;
    meterId: number;
    submittedAt?: Date;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql reads-by-account-id.query.sql: reads-by-account-id.query.sql 1`] = `
"export interface Params {
    id?: number;
    offset?: string;
    perPage?: string;
}
export interface Result {
    id: number;
    msn: string;
    meterId: number;
    dateOn: Date;
    value: number;
    type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    reason?: string;
    createdAt: Date;
    updatedAt?: Date;
    /**
     * sql-fit: MeterReading.MeterReadingId
     */
    sourceSystemId?: number;
    mpan: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql recordsets.query.sql: recordsets.query.sql 1`] = `
"export interface Params {
}
export interface Result {
    id: number;
    exportType: string;
    energy: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Load Files Should convert complex sql tariff-rates.sql: tariff-rates.sql 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;
